include/zephyr/sys/ref.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_REF_H_
+#define ZEPHYR_INCLUDE_SYS_REF_H_
+
+#include <limits.h>
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup sys_ref_apis Reference count APIs
+ * @{
+ */
+
+/* atomics are usually `long`. If they are larger, that's fine because
+ * we'll just artificially limit them . If they were smaller, we'd be unable
+ * to check if an overflow happened. This wouldn't be necessary if the atomic
+ * API told us the max value of an atomic.
+ */
+BUILD_ASSERT(sizeof(atomic_val_t) >= sizeof(long));
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when all references were released. This can also be used on SMP systems.
+ */
+struct sys_ref {
+	atomic_t refs;
+};
+
+/** @brief Initialize sys_ref at compile-time. */
+#define SYS_REF_INIT()                                                                             \
+	{                                                                                          \
+		.refs = ATOMIC_INIT(1),                                                            \
+	}
+
+/** @brief Initialize sys_ref at runtime. */
+static inline void sys_ref_init(struct sys_ref *const ref)
+{
+	*ref = (struct sys_ref){
+		.refs = ATOMIC_INIT(1),
+	};
+}
+
+/**
+ * @brief Create a new reference to `ref`.
+ *
+ * @return The new reference.
```

Comment:
Sort of a weird API.  It's not a new reference, it's a pointer to the same reference struct.  Why bother?

Reply:
You are right. I basically just do what everyone else is doing(net_buf_ref/unref, bt_conn_ref/unref). I didn't find any documentation as for why they do it that way but to me it looks like they are amulating what other programming languages like rust do. In those languages they actually return a new reference object(which holds an internal pointer to the dynamically allocated value). I guess the fact that in C, the returned pointer is the same as the original one is simply an implementation detail that people should not rely on for this kind of API to make sense.